WordPress/Tests/WP/I18nUnitTest.1.inc

```diff
@@ -183,5 +183,34 @@ _n_noop($singular); // Bad x 3.
 // This test is needed to verify that the missing plural argument above does not cause an internal error, stopping the run.
 _n_noop( 'I have %d cat.', 'I have %d cats.' ); // Bad.
 
+// HTML wrapped strings.
+__( '<address>123 Fake Street</address>', 'my-slug' ); // Bad, string shouldn't be wrapped in HTML.
+__( 'I live at <address>123 Fake Street</address>', 'my-slug' ); // Okay, no consensus on HTML tags within strings.
+__( '<address>123 Fake Street</address> is my home', 'my-slug' ); // Okay, no consensus on HTML tags within strings.
+__( '<span>Text</span> More text <span>Text</span>', 'my-slug' ); // Good, we're not wrapping
+__( '<div class="my-class">Translatable content</div>', 'my-slug' ); // Bad
+__( '<option id="%1$s" value="%2$s">Translatable option name</option>', 'my-slug' ); // Wrapping is okay, since there are placeholders
+
+// Strings containing malformed XML, to make sure we're able to parse them. No warnings should be thrown.
+__( '<div>text to translate', 'my-slug' );
+__( 'text to <div>translate', 'my-slug' );
+__( 'text < to translate', 'my-slug' );
+__( 'text to > translate', 'my-slug' );
+__( '<div data-type=">foo">my address</div>', 'my-slug' );
+__( '<div>text to translate<div>', 'my-slug' );
+__( '<>text to translate</>', 'my-slug' );
+__( '<>text to translate<>', 'my-slug' );
+__( '<div><div>', 'my-slug' );
+__( '<<>>text to translate<<>', 'my-slug' );
+__( '<<>>123</<>', 'my-slug' );
+
+// Strings wrapped in '<a href="...">...</a>'. These get a different warning, as the link target might actually be localized.
+// phpcs:disable WordPress.WP.I18n.NoHtmlWrappedStrings
+__( '<a href="http://wordpress.org">WordPress</a>', 'my-slug' ); // Bad, strings shouldn't be wrapped in <a href="...">...</a>.
```

Comment:
This is the one that would concern me. By adding a warning, we're effectively recommending that URLs shouldn't be localized, especially when the comments say that this is "bad".

I'd rather see this one as good, and the placeholder test below possibly as bad.

Reply:
See [this comment](https://github.com/WordPress/WordPress-Coding-Standards/pull/1856#discussion_r374180209), and [the one below](https://github.com/WordPress/WordPress-Coding-Standards/pull/1856#discussion_r374186666).

I ended up adding a separate warning for this case (`NoAHrefWrappedStrings` vs `NoHtmlWrappedStrings`) so that users would be able to separately disable the `<a href=".">...</a>` rule (while retaining the more 'generic' one about wrapping strings in HTML tags), to effectively get the behavior that you're describing. Does that sounds okay?

Reply:
I had read that, and Juliette has the same concerns that I do.

My concern, is that the person writing the code, may not _know_ of any localised link, but is being a good citizen by making the URL available for translation. Yet, the code in this PR is penalising them for that.

If instead of: 

```php
__( '<a href="http://wordpress.org">WordPress</a>', 'my-slug' );
```

It was:

```php
__( '<a href="http://wordpress.org">WordPress</a> Rocks!', 'my-slug' );
```

then this sniff would have no issue with the hard-coded URL.

I'm all for the rest, and I know this currently has a different violation code, but right now, I'd rather let any strings wrapped in `<a ...>...</a>` to not be flagged, regardless of attribute values or placeholders, and then maybe tighten it up in the future if needed.

Reply:
> This is the one that would concern me. By adding a warning, we're effectively recommending that URLs shouldn't be localized, especially when the comments say that this is "bad".

This is incorrect. The recommended approach would be this: `<a href="<?php echo esc_url( __( 'https://wordpress.org/ ' ) ); ?>"><?php echo __( 'WordPress Rocks' ); ?></a>`, and not "not localizing" the URL.

In the approach you recommend (including the `<a>` tag incl the `href` value), translation systems like GlotPress (on translate.wordpress.org) will actually throw a warning if you want to submit a translation with a changed tag (and changing the URL in a tag is changing the tag), so I'd consider this discouraged as well.

This is why I think it's preferrable to wrap the URL separately in gettext and have the sniff advise so (i.e. change the message `Strings should not be wrapped in <a href="...">...</a>` to something like `Strings should not be wrapped in <a href="...">...</a>, please wrap the URL separately`).

Reply:
> The recommended approach would be this: `<a href="<?php echo esc_url( __( 'https://wordpress.org/' ) ); ?>"><?php _e( 'WordPress Rocks' ); ?></a>`, and not "not localizing" the URL.

Your example doesn't match what I put. In mine, `WordPress` was linked, while `Rocks` was not. By following your approach, the string would be split.

https://developer.wordpress.org/apis/handbook/internationalization/internationalization-guidelines/#best-practices-for-writing-strings says "Do not leave URLs for translation, unless they could have a version in another language.".

In the case of my example, I'd want to switch it to `https://en-gb.wordpress.org`.

An examples where it wouldn't be switched, would be when it points to an internal settings page.

At this stage, I wouldn't want to get into analysing the URL itself to decide whether to flag it or not, and so, would still like to see this not flagged for this first pass. With more discussion, and consensus, maybe a variant of this check for a href could be added.

Reply:
> Your example doesn't match what I put. In mine, WordPress was linked, while Rocks was not. By following your approach, the string would be split.

This was for the sake of an argument. The whole point of this PR is to encourage leaving out as much as HTML possbile from Gettext string. Your string of argument is distracting from this approach.

> Correct, along with warnings for other validation failures as well - that's just a way to highlight it to Translation Editors, rather than let potentially spam URLs to make it through to production translation sets. It's not to discourage valid localised URLs from being submitted.

We can disagree about the reasons behind the validation failure. In general your approach encourages translators to change HTML tags which in my opinion should not be their business and they should actually be protected. It's easy to introduce invalid HTML and just dismiss the warning. So actually the fact that you can override warnings in GlotPress is not helpful in this case.

Reply:
> The whole point of this PR is to encourage leaving out as much as HTML possbile from Gettext string.

Your definition of "as possible" and mine clearly differ. I'll refer to my [previous comment](https://github.com/WordPress/WordPress-Coding-Standards/issues/1419#issuecomment-403797241) on the issue as for why.

> In general your approach encourages translators to change HTML tags which in my opinion should not be their business and they should actually be protected.

_Your opinion_ disagrees with the `gettext` manual, WP best practices document, and one or two handbooks.

Besides that, the prior art here is that the translators using GlotPress have been handling (wrapping) markup in strings from developers who didn't know better for a number of years now. That, on its own, isn't necessarily reason enough to continue to allow it, but I can't see what major issue has happened that would warrant the overly strict set of checks that you want to see imposed in the first pass.

As I've said, I'm happy to continue this discussion for wrapping anchor links, but by moving it outside of this PR, we can unblock the majority of the benefit that the rest of it brings.